pkg/pulseserver/status_dispatcher.go

```diff
@@ -0,0 +1,310 @@
+// 2. Periodically checks the active maestro instances and consumers to maintain the consistent hash ring.
+// 3. Signals resync for newly added consumers for the current instance.
+// 4. Starts the status resync workers to process status resync requests.
+// 5. Subscribes to resource status update messages.
+//
+// The function runs until the provided context is canceled or an error occurs.
+func (s *pulseServerWithStatusDispatcher) Start(ctx context.Context) error {
+	logger := logger.NewOCMLogger(ctx)
+	instance := &api.ServerInstance{
+		Meta: api.Meta{
+			ID:        s.instanceID,
+			UpdatedAt: time.Now(),
+		},
+	}
+	_, err := s.instanceDao.UpSert(ctx, instance)
+	if err != nil {
+		return fmt.Errorf("unable to create maestro instance: %s", err.Error())
+	}
+
+	instances, err := s.instanceDao.All(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to list maestro instances: %s", err.Error())
+	}
+
+	// initialize the consistent hash ring
+	members := []consistent.Member{}
+	for _, instance := range instances {
+		// Instances not pulsing within the last three check intervals are considered as inactive.
+		if instance.UpdatedAt.After(time.Now().Add(time.Duration(int64(-3*time.Second) * s.checkInterval))) {
+			members = append(members, instance)
+		}
+	}
+
+	s.consistent = consistent.New(members, consistent.Config{
+		PartitionCount:    30,   // consumer IDs are distributed among partitions, select a big PartitionCount for more consumers.
+		ReplicationFactor: 20,   // the numbers for maestro instances to be replicated on consistent hash ring.
+		Load:              1.25, // Load is used to calculate average load, 1.25 is reasonable for most cases.
+		Hasher:            hasher{},
+	})
+	logger.V(4).Infof("Initialized consistent hash ring with members: %v", members)
+
+	consumers, err := s.consumerDao.All(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to list consumers: %s", err.Error())
+	}
+
+	toAddConsumers := []string{}
+	// initialize the consumer set for current instance
+	for _, consumer := range consumers {
+		instanceID := s.consistent.LocateKey([]byte(consumer.ID)).String()
+		if instanceID == s.instanceID {
+			// new consumer added to the current instance, need to resync resource status updates for this consumer
+			logger.V(4).Infof("Adding new consumer %s to consumer set for instance %s", consumer.ID, s.instanceID)
+			toAddConsumers = append(toAddConsumers, consumer.ID)
+			s.workQueue.Add(consumer.ID)
+		}
+	}
+	_ = s.consumerSet.Append(toAddConsumers...)
+	logger.V(4).Infof("Initialized consumers %s for current instance %s", s.consumerSet.String(), s.instanceID)
+
+	// start the status resync workers
+	go s.startStatusResyncWorkers(ctx)
+
+	// start subscribing to resource status update messages.
+	s.startSubscription(ctx)
+
+	pulseTicker := time.NewTicker(time.Duration(s.pulseInterval) * time.Second)
+	checkTicker := time.NewTicker(time.Duration(s.checkInterval) * time.Second)
```

Comment:
is this part the same code? I think should have one pulseServer implementation without code duplication, what happens when instance is down or consumer is added can be defined by a handler interface.

Reply:
Thanks, now the `pulse server` only pulses and checks instances.
Also `pulse server` has a `Dispatcher` (which is an interface) that has different implementation for shared subscription type and hashing type.
The `Dispatcher` has handler functions(OnInstanceUp, OnInstanceDown) which will be called by pulse server.